### PR TITLE
Refactoring of Evaluation and adding of evaluate command

### PR DESCRIPTION
**Pull request by bigximik**
# ✨ Description

Creates **Evaluator** abstraction so additional evaluators beyond **Loss** can be added.

Adds an `evaluate` command that accepts the same training config and enables evaluation on the last checkpoint.

Includes some fixes.

### Example: specifying multiple **LossEvaluators**
```yaml
training:
  evaluators:
    the_stack:
      interval: 50
      evaluator:
        type: loss
        iterations: 25
        dataset_name: the_stack
    fineweb:
      interval: 100
      evaluator:
        type: loss
        iterations: 15
        dataset_name: fineweb
data:
  datasets:
    the_stack:
      type: file
      path: path/to/validation_the_stack_dataset.yaml
    fineweb:
      type: file
      path: path/to/validation_fineweb_dataset.yaml
```

## 🔍 Type of change

Select all that apply:

- [x] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [x] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [x] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [x] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [x] 🎉 The functionality is complete, and I have tested the changes.
- [x] 📝 I have updated the documentation if needed.
- [x] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [x] 🧩 I have commented my code, especially in hard-to-understand areas.

### Testing

- [x] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes.
- [x] 🚦 I have tested these changes on GPUs and verified training stability.

### CODE REVIEW

File: fast_llm/engine/training/config.py [hunk 1]
```diff
@@ -151,22 +152,41 @@ class WandbConfig(Config):
...
+
+    def get_run_count(self, training_iterations: int, extra_evaluations: int = 0):
+        # Number of completed evaluation runs
+        return (
+            (self.run_interval.get_count(training_iterations) + extra_evaluations)
+            if self.run_interval.enabled()
+            else 0
+        )
+
+    def get_evaluator(
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/training/config.py:167]: I explained in details https://github.com/ServiceNow/Fast-LLM/pull/222#discussion_r2083894400 and https://github.com/ServiceNow/Fast-LLM/pull/222#discussion_r2083904774 but shortly it is to maintain proper incapsulating

File: fast_llm/data/data/gpt/config.py [hunk 2]
```diff
@@ -85,5 +85,6 @@ def _from_dict(
                     assert rename not in default["datasets"]
                     default["datasets"][rename] = default["datasets"].pop(phase.value)
 
-        cls._handle_renamed_field(default, "validation", ("evaluations", "validation"))
+        # TODO: why it is here?
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/data/data/gpt/config.py:88]: if you also have no idea why it is here i will remove as it looks to be in the wrong place

[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/data/data/gpt/config.py:88]: Also maybe we need to move dataset definitions to `EvaluatorLossConfig` as it seems only it will be using it?

File: fast_llm/engine/training/config.py [hunk 3]
```diff
@@ -324,7 +344,8 @@ def _from_dict(
         flat: bool = False,
     ) -> typing.Self:
         # TODO v0.x: Remove backward compatibility.
-        cls._handle_renamed_field(default, "validation", ("evaluations", "validation"))
+        cls._handle_renamed_field(default, "validation", ("evaluators", "validation"))
+        cls._handle_renamed_field(default, "evaluations", ("evaluators"))
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:348]: We did some back and forth with this, I don't remember the final choice but either one is OK

File: fast_llm/engine/training/config.py [hunk 4]
```diff
@@ -151,22 +152,41 @@ class WandbConfig(Config):
...
-        " Setting to None will disable evaluation."
-    )
-    offset = FieldUpdate(desc="Offset for the first evaluation phase.")
-    iterations: int | None = Field(
-        default=None,
-        desc="Number of iterations for each evaluation phase. Setting to None will disable.",
-        hint=FieldHint.feature,
-        valid=skip_valid_if_none(check_field(Assert.gt, 0)),
-    )
+class TrainingEvaluatorConfig(EvaluatorConfigBase):
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:155]: Multiple inheritance is only needed because of the `EvaluatorConfigBase` mixin, which is arguably not even necessary. I'd rather prioritize better usage (configs) over marginally simpler code.

[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/training/config.py:155]: OK, I will use multiple inheritance but will retain `EvaluatorConfigBase` so I don't need to rewrite `EvaluatorRunner`  then we will create a separate config for evaluate command in addition for it of accepting training config. We need to discuss what it should look like, so I’ve created an issue for it #285.

File: tests/test_gpt_loss.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    iter_losses, _, _ = runner.run_step(
+        iter((((inputs_, kwargs),),)), schedule, iteration=iteration, preprocessed=True
+    )
+
+    return iter_losses
+
+
+# @pytest.mark.extra_slow
+@requires_cuda
+def test_loss_validation_vs_inference(model_and_tokenizer):
```
Comment (jlamypoirier) [inline tests/test_gpt_loss.py:112]: I don't think this test is worth it, it's kind of trivial.
